### DISCUSSION TITLE
tailwindcss 동적 스타일링 관련 문제

### DISCUSSION THREAD

**Discussion by seongminn** [prgrms-fe-devcourse/FEDC5_Owhat_Byunghyun#39] [Announcements]
안녕하세요 여러분. discussions 탭을 처음 사용하게 되었네요ㅎㅎ

다름이 아니라 [Group 컴포넌트](https://github.com/prgrms-fe-devcourse/FEDC5_Owhat_Byunghyun/pull/34)를 개발하다가 동적 스타일링과 관련하여 문제가 생겨 의논하고자 글을 남깁니다.

테일윈드는 빌드 시 테일윈드 클래스네임을 모두 css 코드로 파싱하는데, 제가 작성한 코드는 런타임에 동적으로 클래스네임을 생성하고 있기 때문에 제대로 작동하지 않습니다. 그래서 보시는 것처럼 style 속성을 활용하여 스타일을 넣어주었어요. 근데 이게 `v2`에서는 권장하는 방식이었지만 `v3`로 업데이트 되면서 공식문서에서도 찾아볼 수 없는 거 보니 권장하는 방식은 아닌가봐요. 

그렇다고 딱 정해진 값으로만 spacing을 조절하도록 하는 건 너무 제한적이지 않을까 싶은 생각이 자꾸만 들어 같이 이야기해서 방법을 정하면 좋을 것 같습니다. 제가 생각한 방법은 다음과 같아요.

1. style 속성을 사용하여 동적으로 값을 변경할 수 있도록 한다. (제가 처음 작성한 코드)
  - 장점: 별다른 코드 추가 없이 간편하게 스타일을 조정할 수 있다.
  - 단점: 공식문서에서 권장하고 있지는 않다. 다만, 주의하되 꼭 필요하다면 사용하는 것이 문제를 일으킬 것 같지는 않음(GPT의 말씀)
2. `sm`, `md`, `lg` 등 우리가 미리 정의한 variant로만 사용한다.
  - 장점: 공식문서에서 권장하는 내용
  - 단점: 배리에이션이 너무나 한정적임
3. 테일윈드 config type을 가져올 수 있는 라이브러리 설치
  - 장점: props로 테일윈드 클래스네임 자체를 넘겨 받을 수 있음(ex. `<Component color="bg-blue-500" spacing="gap-5" />`)
  - 단점: weekly download 수도 적고, last published가 4개월 전인 만큼 신뢰도가 높지 않고 위험 부담이 큼

꼭 한 번씩 읽어보시고 의견 남겨주세요! @Heeeera @oridori2705 @mibu119 @HY219

**Comment by oridori2705** [comment]
저는 style속성 사용하는 게 좋을 것 같아요
동적으로 처리해야할 경우가 많은데 테일윈드에서 권장한다는 얘기는 없어도
퀄리티 높은 결과물을 위해서 사용하면 좋을 것 같습니다.
만약 이 부분이 면접에서 
**'style속성을 사용하는 로직이 테일윈드에서 권장하는 기능인지? 동적으로 처리할 떄 이 방법을 사용한 이유가 있는지?'** 한다면 
**'저희는 동적으로 처리해야하는 경우가 많았고 먼저 퀄리티 높은 결과물을 우선시해야 했기때문에 이 같은 방법을 사용했습니다.'**
라고 생각하면 될 것 같아요!

**Comment by Heeeera** [comment]
저도 Spacer 컴포넌트를 구현하면서 같은 문제가 있었어요!

variants에 모든 값을 일일이 등록하자니 번거로운 것 같고, 일부 값만 사용한다면 너무 제한적이라는 생각이 들더라구요.
공식 문서가 권장하는 방식을 따르는 것이 좋을테고 Tailwind를 사용하면서 인라인 스타일을 쓰는게 내키진 않지만.. `gap`, `margin`처럼 일부 값만 특정해내기 어려운 스타일은 style prop을 사용하는 것이 좋다고 생각합니다!

**Reply by mibu119** [reply]
저두 희라님 말씀에 동의합니다~

**Comment by mibu119** [comment]
저도 제가 생각하기에 정해진 값으로만 하는 것은 제한적인 느낌이 들 것 같아서 자유도를 넓히려면 지금 하신대로 하는게 맞다고 생각합니다. 공식문서에서 제안한 방법이 아니라는 것은 혹시 성능의 문제가 생겨서 고민이 되시는 걸까요? 그게 아니라면 앞서 준혁님이 말씀하신대로 생각하면 될 것 같아요
